File: packages/core-php/src/WebAuthn.php
<?php

declare(strict_types=1);

namespace OpenPasskey;

class WebAuthn
{
    public static function verifyRegistration(
        string $rpId,
        string $expectedChallenge,
        string $expectedOrigin,
        string $clientDataJSON,
        string $attestationObject,
        bool $requireUserVerification = false,
    ): RegistrationResult {
        $clientDataJSONRaw = ClientData::verify(
            $clientDataJSON,
            'webauthn.create',
            $expectedChallenge,
            $expectedOrigin,
        );

        $att = self::decodeAttestationObject($attestationObject);

        $pad = AuthData::parse($att['authData'], true);

        AuthData::verifyRpIdHash($pad->rpIdHash, $rpId);
        AuthData::checkFlags($pad->flags, $requireUserVerification);

        if ($att['fmt'] === 'packed') {
            Packed::verify($att['attStmt'], $att['authData'], $clientDataJSONRaw, $pad->credentialKey);
        }

        return new RegistrationResult(
            credentialId: Base64Url::encode($pad->credentialId),
            publicKeyCose: Base64Url::encode($pad->credentialKey),
            signCount: $pad->signCount,
            rpIdHash: Base64Url::encode($pad->rpIdHash),
            flags: $pad->flags,
            backupEligible: ($pad->flags & 0x08) !== 0,
            backupState: ($pad->flags & 0x10) !== 0,
            attestationFormat: $att['fmt'],
            attestationX5C: $att['attStmt']['x5c'] ?? null,
        );
    }

    public static function verifyAuthentication(
        string $rpId,
        string $expectedChallenge,
        string $expectedOrigin,
        string $storedPublicKeyCose,
        int $storedSignCount,
        string $clientDataJSON,
        string $authenticatorData,
        string $signature,
        bool $requireUserVerification = false,
    ): AuthenticationResult {
        $clientDataJSONRaw = ClientData::verify(
            $clientDataJSON,
            'webauthn.get',
            $expectedChallenge,
            $expectedOrigin,
        );

        $authDataRaw = Base64Url::decode($authenticatorData);

        $pad = AuthData::parse($authDataRaw, false);

        AuthData::verifyRpIdHash($pad->rpIdHash, $rpId);
        AuthData::checkFlags($pad->flags, $requireUserVerification);

        $sigBytes = Base64Url::decode($signature);

        Signature::verify($storedPublicKeyCose, $authDataRaw, $clientDataJSONRaw, $sigBytes);

        if ($storedSignCount > 0 && $pad->signCount <= $storedSignCount) {
            throw new WebAuthnException('sign_count_rollback');
        }

        return new AuthenticationResult(
            signCount: $pad->signCount,
            flags: $pad->flags,
            backupEligible: ($pad->flags & 0x08) !== 0,
            backupState: ($pad->flags & 0x10) !== 0,
        );
    }

    private static function decodeAttestationObject(string $attObjB64): array
    {
        $raw = Base64Url::decode($attObjB64);
        $map = CborDecoder::decode($raw);

        $fmt = $map['fmt'] ?? null;
        $authData = $map['authData'] ?? null;

        if ($fmt === null || $authData === null) {
            throw new WebAuthnException('invalid_attestation_statement', 'Missing fmt or authData');
        }

        if ($fmt === 'none') {
            return ['fmt' => 'none', 'authData' => $authData, 'attStmt' => []];
        }

        if ($fmt === 'packed') {
            $attStmt = $map['attStmt'] ?? [];
            if (!isset($attStmt['alg']) || !isset($attStmt['sig'])) {
                throw new WebAuthnException('invalid_attestation_statement', 'Missing alg or sig in attStmt');
            }
            return ['fmt' => 'packed', 'authData' => $authData, 'attStmt' => $attStmt];
        }

        throw new WebAuthnException('unsupported_attestation_format', "Unsupported format: {$fmt}");
    }
}

File: packages/core-ruby/lib/open_passkey/webauthn.rb
# frozen_string_literal: true

require "openssl"

module OpenPasskey
  module WebAuthn
    class << self
      def verify_registration(rp_id:, expected_challenge:, expected_origin:,
                              client_data_json:, attestation_object:,
                              require_user_verification: false)
        client_data_json_raw = ClientData.verify(
          client_data_json, "webauthn.create", expected_challenge, expected_origin
        )

        att = decode_attestation_object(attestation_object)

        pad = AuthData.parse(att[:auth_data], true)

        AuthData.verify_rp_id_hash(pad.rp_id_hash, rp_id)
        AuthData.check_flags(pad.flags, require_user_verification)

        if att[:fmt] == "packed"
          Packed.verify(att[:att_stmt], att[:auth_data], client_data_json_raw, pad.credential_key)
        end

        RegistrationResult.new(
          credential_id: Base64Url.encode(pad.credential_id),
          public_key_cose: Base64Url.encode(pad.credential_key),
          sign_count: pad.sign_count,
          rp_id_hash: Base64Url.encode(pad.rp_id_hash),
          flags: pad.flags,
          backup_eligible: (pad.flags & 0x08) != 0,
          backup_state: (pad.flags & 0x10) != 0,
          attestation_format: att[:fmt],
          attestation_x5c: att[:att_stmt]["x5c"]
        )
      end

      def verify_authentication(rp_id:, expected_challenge:, expected_origin:,
                                stored_public_key_cose:, stored_sign_count:,
                                client_data_json:, authenticator_data:, signature:,
                                require_user_verification: false)
        client_data_json_raw = ClientData.verify(
          client_data_json, "webauthn.get", expected_challenge, expected_origin
        )

        auth_data_raw = Base64Url.decode(authenticator_data)

        pad = AuthData.parse(auth_data_raw, false)

        AuthData.verify_rp_id_hash(pad.rp_id_hash, rp_id)
        AuthData.check_flags(pad.flags, require_user_verification)

        sig_bytes = Base64Url.decode(signature)

        Signature.verify(stored_public_key_cose, auth_data_raw, client_data_json_raw, sig_bytes)

        if stored_sign_count > 0 && pad.sign_count <= stored_sign_count
          raise WebAuthnError.new("sign_count_rollback")
        end

        AuthenticationResult.new(
          sign_count: pad.sign_count,
          flags: pad.flags,
          backup_eligible: (pad.flags & 0x08) != 0,
          backup_state: (pad.flags & 0x10) != 0
        )
      end

      private

      def decode_attestation_object(att_obj_b64)
        raw = Base64Url.decode(att_obj_b64)
        map = CborDecoder.decode(raw)

        fmt = map["fmt"]
        auth_data = map["authData"]

        if fmt.nil? || auth_data.nil?
          raise WebAuthnError.new("invalid_attestation_statement", "Missing fmt or authData")
        end

        if fmt == "none"
          return { fmt: "none", auth_data: auth_data, att_stmt: {} }
        end

        if fmt == "packed"
          att_stmt = map["attStmt"] || {}
          unless att_stmt["alg"] && att_stmt["sig"]
            raise WebAuthnError.new("invalid_attestation_statement", "Missing alg or sig in attStmt")
          end
          return { fmt: "packed", auth_data: auth_data, att_stmt: att_stmt }
        end

        raise WebAuthnError.new("unsupported_attestation_format", "Unsupported format: #{fmt}")
      end
    end
  end
end
